### ISSUE TITLE
Possible D7 upgrade regression caused by user/1 path alias update hook

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#5497]
## Description of the bug

When upgrading from D7, following exceptions pop up:

```
PDOException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'auto' in 'field list': INSERT INTO {url_alias} (`pid`, `source`, `alias`, `langcode`, `auto`) VALUES (default, :db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2, :db_insert_placeholder_3); Array ( [:db_insert_placeholder_0] => user/1 [:db_insert_placeholder_1] => accounts/superadmin [:db_insert_placeholder_2] => und [:db_insert_placeholder_3] => 1 ) in backdrop_write_record() (line 8288 of /.../core/includes/common.inc).

EntityStorageException: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'auto' in 'field list' in EntityDatabaseStorageController->save() (line 786 of ...core/modules/entity/entity.controller.inc).
```
Possibly introduced with Issue #1417: User URL alias not applied to admin user out of the box.

## Steps To Reproduce

Upgrade from a Drupal 7

## Actual behavior

Exception. Note that the alias for user/1 does get created (in a second update run).

## Expected behavior

No exception

## Additional information

Comparing the D7 and B1 database structure, table url_alias has a column auto, but the Drupal table has not.
That column gets added in `function system_update_1039()`.
If we let function user_update_1021() depend on the system update hook, that should be fine, because then the column exists.

See also: [hook_update_dependencies](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21system.api.php/function/hook_update_dependencies/1)

**Comment by indigoxela** [comment]
A [PR](https://github.com/backdrop/backdrop/pull/3948) is available for testing and review.

Note that testing requires a local instance, and a D7 database to run the upgrade from - a little inconvenient, I know.

For my tests I set up a relatively simple Drupal install with a handful of modules (admin_menu, ctools, views, date, entity, transliteration, pathauto, ckeditor - the usual suspects).

**Comment by indigoxela** [comment]
Wow, I almost missed that, the user module implements "system_update_dependencies()" in user.install? Also fixed. :wink:

**Comment by klonos** [comment]
Nice ...I wasn't aware of `hook_update_dependencies()` (never had to use it directly or change any code that requires/uses it). I'll test soon.

**Comment by argiepiano** [comment]
I've tried testing this but I'm unable to reproduce the error. Is there a step missing in the description? 

I started with a plain Drupal 7 site, with just a couple of contrib modules enabled (admin_menu_toolbar, module_filter).  I followed the usual migration steps. PHP 7.4 and Backdrop 1.22.x-dev. Any clues?

**Comment by argiepiano** [comment]
One more piece of information about the lack of error in my migration: the url_alias table is completely empty - meaning that no aliases were created when migrating. When visiting `admin/config/people/settings` I see that there is no pattern set for user aliases. 

That's also the case for content - no patterns are set. 

I think, in order to see this error, you need to start with a D7 site that either used pathauto, and/or had already created all those aliases before the migration. @indigoxela , can you confirm this? I will test that later and will test the patch as well.

**Comment by argiepiano** [comment]
I'm still unable to reproduce this error. I have installed Pathauto in my D7 and "bulk-generated" aliases for content and users. I then proceeded to migrate the site to Backdrop, without any errors. I checked the alias patterns - they are all there. All aliases were created correctly. 

One possible explanation is that, in my system, the user_update_1021 is normally running after system_update_1039, and therefore there are no issues, but in someone else's installation that's not the case (though I still find it very strange that the user_update hooks would run before system_update ones).

Because of this, I really can't test this PR, since I can't reproduce the error.

**Comment by indigoxela** [comment]
> Because of this, I really can't test this PR, since I can't reproduce the error. ... Any clues?

Yes. It's a matter of order. The order of update hooks can _by accident_ be correct - in that case (table already altered) no problem appears. But as the update dependency currently isn't defined, it can happen that the user update hook runs *before* the system update hook, leading to the exception. Only a dependency definition prevents that. Especially when a lot more than just the core update hooks need to run. That's why it's useful to create a more realistic D7 setup with several contribs in place.

**Comment by BWPanda** [comment]
From a coding point of view the PR looks good to me, though I can't test or confirm that it actually works.

**Comment by quicksketch** [comment]
I ran into this same error on a Drupal 7 upgrade after weeks of the upgrade working fine (I run it once a week), I added a new update hook and the order started coming in incorrectly because of the lack of the declared dependency. I reached the identical solution separately before I found this issue. Thanks @indigoxela for finding and filing this PR!

And thanks for changing `system_update_dependencies()` to `user_update_dependencies()` in `user.install` as well.
